modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Resources.scala

```diff
@@ -116,45 +127,112 @@ object Resources {
   private def mkTypeSink[F[_]: Concurrent](
     projectId: String,
     topic: String,
-    maxConcurrency: Int
+    sinkSettingsTypes: SinkSettings.Types
   ): Resource[F, Pipe[F, Set[ShreddedType], Unit]] =
-    mkProducer[F, Set[ShreddedType]](projectId, topic, 4L, 200.millis).map { p =>
-      _.parEvalMapUnordered(maxConcurrency) { types =>
+    mkProducer[F, Set[ShreddedType]](
+      projectId,
+      topic,
+      sinkSettingsTypes.producerBatchSize,
+      sinkSettingsTypes.producerDelayThreshold
+    ).map { p =>
+      _.parEvalMapUnordered(sinkSettingsTypes.sinkConcurrency) { types =>
         p.produce(types).void
       }
     }
 
   /** Write good events through a BigQuery sink and pipe observed types through a Pub/Sub sink.
-    *  Acks the event.
+    * Acks the events.
+    *
+    * The input Stream of events is transformed into a Stream of batches of events (List[StreamLoaderRow]).
+    * Each batch respects the Streaming APIs recommended number of records per request.
+    * Each BQ write request also has a size limit of 10MB. In case a batch's total size exceeds this limit, we use
+    * as many events as possible (while staying below it) in the write request, and re-insert the remaining events
+    * back into the original Stream of events.
+    *
+    * @param bufferOverflowQueue A Queue in which put events that are left over from a batch after taking as many as
+    *                            possible without breaching the streaming inserts request size limit
     */
   private def mkGoodSink[F[_]: Concurrent: ContextShift: Timer](
     blocker: Blocker,
     good: Output.BigQuery,
     bigQuery: BigQuery,
     producer: PubsubProducer[F, FailedInsert],
     metrics: Metrics[F],
-    typesSink: Pipe[F, Set[ShreddedType], Unit]
+    typeSink: Pipe[F, Set[ShreddedType], Unit],
+    bufferOverflowQueue: Queue[F, StreamLoaderRow[F]],
+    sinkSettingsGood: SinkSettings.Good,
+    sinkSettingsTypes: SinkSettings.Types
   ): Pipe[F, StreamLoaderRow[F], Unit] = {
-    val goodPipe: Pipe[F, StreamLoaderRow[F], Unit] = _.parEvalMapUnordered(MaxConcurrency) { slrow =>
-      blocker.blockOn(
-        Bigquery.insert(producer, metrics, slrow.row)(Bigquery.mkInsert(good, bigQuery)) *> slrow.ack
-      )
+    val goodPipe: Pipe[F, StreamLoaderRow[F], Unit] = {
+      _.merge(bufferOverflowQueue.dequeue)
+        .through(batch(sinkSettingsGood.bqWriteRequestThreshold, sinkSettingsGood.bqWriteRequestTimeout))
```

Comment:
By doing the splitting here right after the batching you don't need the queue any more. I'm aware that it can result in requests of 1 record though. Don't know about the statistics of how often this would happen.

If you're up for challenge, the ideal think to do would be to write your own [implementation of groupWithin](https://github.com/typelevel/fs2/blob/v2.5.0/core/shared/src/main/scala/fs2/Stream.scala#L1363-L1462) to take the size into account directly.

Reply:
I don't know what the stats are either, but to me it looks like it would be happening all the time for customers with big events. `splitBy` will have to return not a tuple but a new `List[List[LoaderRow]]` where each inner List will have to fit within the max size and it seems to me that there will often be at least one that will be much smaller than the rest.